core/src/Streamly/Internal/FileSystem/Posix/ReadDir.hsc

```diff
@@ -148,41 +269,76 @@ isMetaDir dname = do
                 then return True
                 else return False
 
-{-# NOINLINE lstatDname #-}
-lstatDname :: PosixPath -> Ptr CChar -> IO (Bool, Bool)
-lstatDname parent dname = do
+data GStatRes
+    = GSIsMetaDir
+    | GSIsRegDir
+    | GSIsNotDir
+    | GSIgnoreError Errno
+
+{-# NOINLINE gstatDname #-}
+gstatDname
+    :: ReadOptions -> PosixPath -> Ptr CChar -> IO GStatRes
+gstatDname conf parent dname = do
     isMeta <- liftIO $ isMetaDir dname
     if isMeta
-    then pure (True, True)
+    then pure GSIsMetaDir
     else do
         -- XXX We can create a pinned array right here since the next call pins
         -- it anyway.
         path <- appendCString parent dname
         Array.asCStringUnsafe (Path.toChunk path) $ \cStr -> do
-            res <- c_lstat_is_directory cStr
+            let sfunc = if _followSymlinks conf then c_stat else c_lstat
+            res <- statWith sfunc cStr
             case res of
-                x | x == 1 -> pure (True, False)
-                x | x == 0 -> pure (False, False)
-                -- XXX Need to check if and how we should handle some errors
-                -- like EACCES.
-                _ -> throwErrno "checkIfDirectory"
+                Left errno ->
+                    if errno == eNOENT
+                    then
+                        if _ignoreNonExistingFiles conf
+                        then pure $ GSIgnoreError errno
+                        else throwErrno (errMsg path)
+                    else if errno == eACCES
+                    then
+                        if _ignoreInAccessibleFiles conf
+                        then pure $ GSIgnoreError errno
+                        else throwErrno (errMsg path)
+                    else if errno == eLOOP
+                    then
+                        if _ignoreSymlinkLoopErrors conf
+                        then pure $ GSIgnoreError errno
+                        else throwErrno (errMsg path)
+                    else throwErrno (errMsg path)
+                Right mode -> pure $
```

Comment:
In general, in if or case statements, better keep the most common case as the first case, in case the compiler generates code assuming that the first branch is more likely.